src/codelabs/async-await/index.md

```diff
@@ -0,0 +1,473 @@
+  * Because `getNewUsername()` can encounter errors, the `changeUsername()` function must __catch and return any errors__ and it must __stringify the error before returning it__.
+* You can use the [toString()]({{site.dart_api}}/stable/dart-core/ArgumentError/toString.html) method to stringify both [Exceptions]({{site.dart_api}}/stable/dart-core/Exception-class.html) and [Errors.]({{site.dart_api}}/stable/dart-core/Error-class.html)
+
+<!-- [//]: https://gist.github.com/858f71f0ad0e70051999bcafa41806a3 -->
+
+<iframe 
+src="https://dartpad.dartlang.org/experimental/embed-new-dart.html?id=858f71f0ad0e70051999bcafa41806a3&theme=dark"
+frameborder="no"
+height="525"
+width="100%" >
+</iframe>
+
+### Exercise: Putting it all together
+
+It's time to practice what you've learned in one final exercise. 
+To simulate asynchronous operations, __this exercise provides the async functions
+`getUsername()` and `logoutUser()`. You don’t need to implement these provided functions.
+You don't need to implement `main()`.
+
+To simulate asynchronous operations, your code will call the following functions, which are 
+provided for you:
+
+{% prettify dart %}
+getUsername() → Future<String> 
+// Returns the name associated with the current user.
+
+logoutUser() → Future<String> 
+// Performs logout of current user and returns the username that was logged out.
+{% endprettify %}
+
+Write the following:
+
+####  Part 1: `addHello()` 
+* Write a function `addHello()` that takes a single String argument 
+<!-- try to avoid should -->
+* `addHello()` should return its String argument surrounded by the text ‘Hello \<string\>'
+
+####  Part 2: `greetUser()` 
+* Write a function `greetUser()` that takes no arguments.
+* To obtain the username, `greetUser()` should call the provided asynchronous function `getUsername()`
+* `greetUser()` should create a greeting for the user by calling `addHello()`, passing it the username, and returning the result. 
+  * For example, if the username is "Jenny", `greetUser()` should create and return the greeting "Hello Jenny"
+
+####  Part 3: `sayGoodbye()` 
+<!-- pull out descriptions of  -->
+* Write a function `sayGoodbye()` that does the following:
+* `sayGoodbye()` takes no arguments.
+* `sayGoodbye()` catches any errors.
+* `sayGoodbye()` calls the provided asynchronous function `logoutUser()`.
+<!-- italicize <result> -->
+* If `logoutUser()` succeeds, `sayGoodbye()` returns the string "\<result\> Thanks, see you next time" where \<result\> is the String value returned by calling `logoutUser()`.
+
+    <!-- [//]: https://gist.github.com/f601d25bc2833c957186e3c6bf71effc -->
+<iframe 
+  src="https://dartpad.dev/experimental/embed-new-dart.html?id=f601d25bc2833c957186e3c6bf71effc&theme=dark"
+  frameborder="no"
+  height="550"
+  width="100%">
+</iframe>
+
+{{ site.alert.info }}
+  Dart can [infer the return type](https://dart.dev/guides/language/sound-dart#type-inference) for you.
```

Comment:
Need a little context here.

Reply:
Some users didn't know about type inference but wound up encountering it one way or another, and in many cases it was helpful to have a link somewhere when it arose. There's no specific context with respect to the content of the codelab, it's just something that came up. It can come up any number of ways, there's no example that implicated it more than others.

Reply:
Even so, you need a brief reference to some code, or else "infer the return type" means nothing.

Reply:
Do you mean something like: 

`Instead of adding Future<String> to your function, Dart can infer the return type for you...`

Reply:
Yes, something like that. But only if `Future<String>` is already visible.

Also, the actor should be clearer. Here's one possibility: 

You might have noticed that the functions in the exercises don't have return types. That's because Dart can infer the return type for you. Omitting return types is fine when you're prototyping, but when you write production code, we recommend that you specify the return type.

Reply:
Okay, makes sense